### PR TITLE
Encryption

### PR DESCRIPTION
**Pull request by rmsepic**
The file Crypto is a set of functions that I made so that we can encrypt any information that we need to send. My tests are currently with the chat functionality, however, they shouldn't be written that way, I was just trying to send the public key in whatever way I could to test it. 

There will be a design doc along with this PR, and I am putting this up more to receive feedback then to actually get this merged to master. It still needs a bit more work.

### CODE REVIEW

File: Seda/Seda/Crypto.swift [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+            kSecAttrTokenID         as String:   kSecAttrTokenIDSecureEnclave, // Store the key in the enclave
+            kSecPrivateKeyAttrs     as String: [
+                kSecAttrIsPermanent     as String: true,   // Store the key in the keychain, so we will be able to get the key when the user logs back into the application
+                kSecAttrApplicationTag  as String : keyTag
+                
+            ]
+        ]
+        
+        var err: Unmanaged<CFError>?
+        guard let privateKey = SecKeyCreateRandomKey(attributes as CFDictionary, &err) else {
```
Comment (steubs) [inline Seda/Seda/Crypto.swift]: I'm getting "Thread 1: EXC_BAD_ACCESS (code=1, address=0x18)" at this line.

Might be related to this: https://stackoverflow.com/questions/12437605/thread-1-exc-bad-access-code-1-address-0x30000008-issue-generated

File: Seda/Seda/Crypto.swift [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+            kSecClass               as String: kSecClassKey,
+            kSecAttrApplicationTag  as String: keyTag,
+            kSecAttrKeyType         as String: keyType,
+            kSecReturnRef           as String: true
+        ]
+        
+        var item: CFTypeRef?
+        let status = SecItemCopyMatching(query as CFDictionary, &item)
+        guard status == errSecSuccess else {
+            print("Crypto.loadPrivKey(): Error in retrieving the private key")
```
Comment (steubs) [inline Seda/Seda/Crypto.swift]: While I get the above error (EXC_BAD_ACCESS), this line is printed to the console.

**Review by steubs — CHANGES_REQUESTED** [review]

**Comment by rmsepic** [comment]
UPDATE 31st May

I have added friending and end-to-end encryption with the chat app. There was a good bit added to the chat app, so DJ if you want I can move it somewhere else. 

Also, the private key is no longer being stored within the enclave of the phone but within the keychain. It works now in the simulator, but there may be issues when trying to switch devices. 

For testing, I used my phone and a computer, and was able to send and recieve a friend request and message myself.

**Comment by rmsepic** [comment]
Another update on the PR, turns out the issue DJ spotted was not with storage but having to do with friending. Currently storing it on the keychain associated with the **device**, so it can be run in the simulator and will work, however, it will store unique private keys on each simulator device. So I would suggest trying to keep different accounts for each device in the simulator and device as to not mix up the keys.

**Review by Dong-Jie-Chen — APPROVED** [review]

**Pull request merged**